File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/preferences/InCarStorage.java
package com.anod.car.home.prefs.preferences;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import com.anod.car.home.model.ShortcutInfo;

import java.util.ArrayList;
import java.util.Locale;

/**
 * @author algavris
 * @date 19/03/2016.
 */
@SuppressLint("CommitPrefEdits")
public class InCarStorage {
    public static final int NOTIFICATION_COMPONENT_NUMBER = 3;
    private static final String MODE_FORCE_STATE = "mode-force-state";
    private static final String NOTIF_COMPONENT = "notif-component-%d";

    public static final String PREF_NAME = "incar";

    public static SharedPreferences getSharedPreferences(Context context) {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }

    public static InCarSharedPreferences load(Context context) {
        if (PrefsMigrate.required(context)) {
            return PrefsMigrate.migrate(context);
        }

        return new InCarSharedPreferences(getSharedPreferences(context));
    }

    public static void saveScreenTimeout(boolean disabled, boolean disableCharging, InCarSharedPreferences prefs) {
        prefs.setDisableScreenTimeout(disabled);
        prefs.setDisableScreenTimeoutCharging(disableCharging);
        prefs.apply();
    }

    public static String getNotifComponentName(int position) {
        return String.format(Locale.US, NOTIF_COMPONENT, position);
    }

    public static void dropNotifShortcut(int position, Context context) {
        String key = getNotifComponentName(position);
        SharedPreferences prefs = getSharedPreferences(context);
        SharedPreferences.Editor editor = prefs.edit();
        editor.remove(key);
        editor.commit();
    }

    public static ArrayList<Long> getNotifComponents(Context context) {
        return getNotifComponents(getSharedPreferences(context));
    }

    public static void saveNotifShortcut(Context context, long shortcutId, int position) {
        saveNotifShortcut(context, getSharedPreferences(context), shortcutId, position);
    }

    static void saveNotifShortcut(Context context, SharedPreferences prefs, long shortcutId, int position) {
        String key = getNotifComponentName(position);
        WidgetStorage.saveShortcutId(context, prefs, shortcutId, key);
    }

    static ArrayList<Long> getNotifComponents(SharedPreferences prefs) {
        ArrayList<Long> ids = new ArrayList<Long>(NOTIFICATION_COMPONENT_NUMBER);
        for (int i = 0; i < NOTIFICATION_COMPONENT_NUMBER; i++) {
            String key = getNotifComponentName(i);
            long id = prefs.getLong(key, ShortcutInfo.NO_ID);
            ids.add(i, id);
        }
        return ids;
    }

}

File: content/src/main/kotlin/info/anodsplace/carwidget/content/preferences/InCarStorage.kt
package info.anodsplace.carwidget.content.preferences

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import info.anodsplace.carwidget.content.db.ShortcutsDatabase
import java.util.*

/**
 * @author algavris
 * @date 19/03/2016.
 */
object InCarStorage {
    const val NOTIFICATION_COMPONENT_NUMBER = 3
    private const val NOTIF_COMPONENT = "notif-component-%d"
    const val PREF_NAME = "incar"

    fun getSharedPreferences(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
    }

    fun load(context: Context): InCarSettings {
        val prefs = getSharedPreferences(context)
        return InCarSettings(prefs)
    }

    private fun getNotifComponentName(position: Int): String {
        return String.format(Locale.US, NOTIF_COMPONENT, position)
    }

    fun isDbMigrated(context: Context): Boolean {
        val prefs = getSharedPreferences(context)
        if (prefs.contains("db_migrated")) {
            return true
        }
        for (i in 0 until NOTIFICATION_COMPONENT_NUMBER) {
            val key = getNotifComponentName(i)
            val id = prefs.getLong(key, WidgetInterface.idUnknown)
            if (id != WidgetInterface.idUnknown) {
                return false
            }
        }
        return true
    }

    @SuppressLint("ApplySharedPref")
    fun launcherComponentsMigrated(context: Context) {
        val prefs = getSharedPreferences(context)
        val edit = prefs.edit()
        for (i in 0 until NOTIFICATION_COMPONENT_NUMBER) {
            val key = getNotifComponentName(i)
            edit.remove(key)
        }
        edit.putBoolean("db_migrated", true)
        edit.commit()
    }

    fun getMigrateIds(context: Context): ArrayList<Long> {
        val prefs = getSharedPreferences(context)
        val ids = ArrayList<Long>(NOTIFICATION_COMPONENT_NUMBER)
        for (i in 0 until NOTIFICATION_COMPONENT_NUMBER) {
            val key = getNotifComponentName(i)
            val id = prefs.getLong(key, WidgetInterface.idUnknown)
            ids.add(i, id)
        }
        return ids
    }
}
